### PR TITLE
Change markers to use headers to serialize internal fields

### LINKED ISSUE #305: Update SideEffect and MutableSideEffect to use headers
Currently side effect markers serialize a structure that contains a serialized payload into data field. This leads to the following JSON as Marker details:

```
 5: MarkerRecorded [191 ms] {
    Details = {
      "id": "c661391c-feaa-3148-bcb8-a2c401953b50",
      "eventId": 5,
      "data": [
        123,
        34,
   ...
        34,
        58,
        110,
        117,
        108,
        108,
        125
      ],
      "accessCount": 0
    };
    DecisionTaskCompletedEventId = 3;
    MarkerName = MutableSideEffect
  },
```
The proposal is to use header field of the Marker to store id, eventId and accessCount and only put data into the Details.

### PR DESCRIPTION
**Pull request by meiliang86**
resolves #305

### CODE REVIEW

File: src/main/java/com/uber/cadence/internal/common/LocalActivityMarkerData.java [hunk 1]
```diff
@@ -85,77 +92,115 @@ public LocalActivityMarkerData build() {
...
+    return header;
+  }
+
+  public static LocalActivityMarkerData fromEventAttributes(
+      MarkerRecordedEventAttributes attributes) {
+    ByteBuffer byteBuffer = attributes.getHeader().getFields().get(LOCAL_ACTIVITY_HEADER_KEY);
+    byte[] bytes = org.apache.thrift.TBaseHelper.byteBufferToByteArray(byteBuffer);
+    LocalActivityMarkerHeader header =
+        JsonDataConverter.getInstance()
+            .fromData(bytes, LocalActivityMarkerHeader.class, LocalActivityMarkerHeader.class);
```
Comment (mfateev) [inline src/main/java/com/uber/cadence/internal/common/LocalActivityMarkerData.java]: We shouldn't hardcode the converter used to serialize/deserialize the header. The converter should be passed from the higher level code, probably as a parameter to fromEventAttributes and getHeader.

File: src/main/java/com/uber/cadence/internal/replay/DecisionsHelper.java [hunk 2]
```diff
@@ -17,41 +17,7 @@
...
-import com.uber.cadence.StartChildWorkflowExecutionDecisionAttributes;
-import com.uber.cadence.StartChildWorkflowExecutionFailedEventAttributes;
-import com.uber.cadence.StartChildWorkflowExecutionInitiatedEventAttributes;
-import com.uber.cadence.StartTimerDecisionAttributes;
-import com.uber.cadence.TaskList;
-import com.uber.cadence.TimerCanceledEventAttributes;
-import com.uber.cadence.TimerFiredEventAttributes;
-import com.uber.cadence.WorkflowExecutionStartedEventAttributes;
-import com.uber.cadence.WorkflowType;
+import com.uber.cadence.*;
```
Comment (mfateev) [inline src/main/java/com/uber/cadence/internal/replay/DecisionsHelper.java]: revert. There is IntelliJ setting to avoid this.

Comment (meiliang86) [reply src/main/java/com/uber/cadence/internal/replay/DecisionsHelper.java]: Done

File: src/main/java/com/uber/cadence/internal/common/LocalActivityMarkerData.java (hunk 1 above)
Comment (meiliang86) [reply src/main/java/com/uber/cadence/internal/common/LocalActivityMarkerData.java]: Done

File: src/main/java/com/uber/cadence/internal/common/LocalActivityMarkerData.java [hunk 3]
```diff
@@ -85,77 +93,114 @@ public LocalActivityMarkerData build() {
...
+    return headers.backoff;
   }
 
   public boolean getIsCancelled() {
-    return isCancelled;
+    return headers.isCancelled;
+  }
+
+  public Header getHeader() {
+    byte[] headerData = JsonDataConverter.getInstance().toData(headers);
```
Comment (mfateev) [inline src/main/java/com/uber/cadence/internal/common/LocalActivityMarkerData.java]: What about here?

File: src/main/java/com/uber/cadence/internal/replay/ClockDecisionContext.java [hunk 4]
```diff
@@ -88,20 +86,23 @@ public void accept(Exception reason) {
...
   private final Lock laTaskLock = new ReentrantLock();
   private final Condition taskCondition = laTaskLock.newCondition();
   private boolean taskCompleted = false;
 
   ClockDecisionContext(
       DecisionsHelper decisions,
       BiFunction<LocalActivityWorker.Task, Duration, Boolean> laTaskPoller,
-      ReplayDecider replayDecider) {
+      ReplayDecider replayDecider,
+      DataConverter dataConverter) {
```
Comment (mfateev) [inline src/main/java/com/uber/cadence/internal/replay/ClockDecisionContext.java]: I believe DataConverter should live in higher level. At this level everything is a byte. Look at getVersion call in this file.

Comment (meiliang86) [reply src/main/java/com/uber/cadence/internal/replay/ClockDecisionContext.java]: The difference is getVersion is from workflow code, so we can access data converter from SyncDecisionContext and pass it down.
Here deserialization needs to happen while we replay event. So we will need to wire data converter into replay decider.

File: src/main/java/com/uber/cadence/internal/common/LocalActivityMarkerData.java (hunk 3 above)
Comment (meiliang86) [reply src/main/java/com/uber/cadence/internal/common/LocalActivityMarkerData.java]: Fixed.

**Pull request merged**